apps/aecore/src/aec_block_key_candidate.erl

```diff
@@ -66,44 +47,38 @@ adjust_target(Block, AdjHeaders) ->
     end.
 
 %% -- Internal functions -----------------------------------------------------
-int_create(BlockHash, Block, FeesInfo) ->
-    case aec_chain:get_block_state(BlockHash) of
-        {ok, Trees} ->
-            int_create(BlockHash, Block, FeesInfo, Trees);
-        error ->
-            {error, block_state_not_found}
-    end.
 
-int_create(BlockHash, Block, FeesInfo, Trees) ->
+int_create(BlockHash, Block) ->
     N = aec_governance:key_blocks_to_check_difficulty_count(),
     case aec_blocks:height(Block) < N of
         true  ->
-            int_create(BlockHash, Block, FeesInfo, Trees, []);
+            int_create(BlockHash, Block, []);
         false ->
             case aec_chain:get_n_generation_headers_backwards_from_hash(BlockHash, N) of
                 {ok, Headers} ->
-                    int_create(BlockHash, Block, FeesInfo, Trees, Headers);
+                    int_create(BlockHash, Block, Headers);
                 error ->
                     {error, headers_for_target_adjustment_not_found}
             end
     end.
 
-int_create(BlockHash, Block, FeesInfo, Trees, AdjChain) ->
+int_create(BlockHash, Block, AdjChain) ->
     case aec_keys:pubkey() of
         {ok, Miner} ->
-            int_create(BlockHash, Block, FeesInfo, Trees, Miner, AdjChain);
+            int_create(BlockHash, Block, Miner, AdjChain);
         {error, _} = Error ->
             Error
     end.
 
-int_create(PrevBlockHash, PrevBlock, FeesInfo, Trees, Miner, AdjChain) ->
-    {ok, Block, BlockInfo} = int_create_block(PrevBlockHash, PrevBlock, FeesInfo, Trees, Miner),
+int_create(PrevBlockHash, PrevBlock, Miner, AdjChain) ->
+    {ok, Trees} = aec_chain_state:calculate_keyblock_state(PrevBlockHash, Miner),
```

Comment:
Aren't we doing the job twice here? We have already performed `aec_chain_state:insert_block` for `PrevBlock`?

Reply:
yeah, that's good it's shared with aec chain state, tho, so easier to refactor

Reply:
You are misreading. The `calculate_keykblock_state` is calculating what the state is if there is a keyblock added on top of the hash. Should be clarified in the code, and possibly have a better name.

Anyway, the job is done once (here) and if the block is mined, it is performed again with the actual block.